internal/venice-common/src/main/java/com/linkedin/venice/helix/HelixReadOnlySchemaRepository.java

```diff
@@ -372,23 +392,37 @@ public SchemaEntry getSupersetOrLatestValueSchema(String storeName) {
 
   @Override
   public Optional<SchemaEntry> getSupersetSchema(String storeName) {
-    schemaLock.readLock().lock();
+    getSchemaLock().readLock().lock();
     try {
       fetchStoreSchemaIfNotInCache(storeName);
-      SchemaData schemaData = schemaMap.get(getZkStoreName(storeName));
+      SchemaData schemaData = getSchemaMap().get(getZkStoreName(storeName));
       if (schemaData == null) {
         throw new VeniceNoStoreException(storeName);
       }
       Optional<Integer> supersetSchemaID = getSupersetSchemaID(storeName);
+      if (supersetSchemaID.isPresent()) {
+        if (schemaData.getValueSchema(supersetSchemaID.get()) == null) {
+          logger.info(
+              "Superset schema ID: {} for store: {} not in cache, will force refresh schema data.",
+              supersetSchemaID.get(),
+              storeName);
+          forceRefreshSchemaData(getStoreRepository().getStore(storeName), schemaData);
+          if (schemaData.getValueSchema(supersetSchemaID.get()) == null) {
+            throw new InvalidVeniceSchemaException(
+                "Cannot find superset schema ID:" + supersetSchemaID.get() + " for store: " + storeName
+                    + " after force refresh.");
+          }
+        }
+      }
       return supersetSchemaID.map(schemaData::getValueSchema);
 
     } finally {
-      schemaLock.readLock().unlock();
+      getSchemaLock().readLock().unlock();
     }
   }
 
   private Optional<Integer> getSupersetSchemaID(String storeName) {
-    Store store = storeRepository.getStoreOrThrow(storeName);
+    Store store = getStoreRepository().getStoreOrThrow(storeName);
```

Comment:
nit: member var can be access directly.ie no need to use test only `getter` method.

Reply:
Hmm, I need to mock the variable as the main one is mocked.